lib/Doctrine/DBAL/SQLParserUtils.php

```diff
@@ -73,6 +73,42 @@ static public function getPlaceholderPositions($statement, $isPositional = true)
         return $paramMap;
     }
 
+    static private function startsWith($haystack, $needle)
+    {
+        $length = strlen($needle);
+        return (substr($haystack, 0, $length) === $needle);
+    }
+
+    static private function endsWith($haystack, $needle)
+    {
+        $length = strlen($needle);
+        if ($length == 0) {
+            return true;
+        }
+
+        return (substr($haystack, -$length) === $needle);
+    }
+
+    static private function isCollectionType($type)
+    {
+        if ($type === Connection::PARAM_INT_ARRAY || $type === Connection::PARAM_STR_ARRAY)
+            return true;
+
+        return static::startsWith($type, '[') && static::endsWith($type, ']');
+    }
+
+    static private function removeCollectionMarker($type)
+    {
+        if (!static::isCollectionType($type))
+            throw SQLParserUtilsException::notCollectionType($type);
+        
+        if (is_numeric($type))
+            return $type - Connection::ARRAY_PARAM_OFFSET;
+
+        // [my_custom_type] -> my_custom_type
```

Comment:
The DBAL project is already dealing with multiple types:
- string based types (looked up with Type::getType())
- PDO based types
- DBAL types (which are mixed in Connection::getBindingInfo() method)

We have to deal with strings without a major and full of BC breaks refacto ; this is how it currently works with Connection::ARRAY_PARAM_OFFSET. 

I understand this part is a major piece of the cake, maybe someone else could come with a better idea? We're already using this on our own project, and this is working well :)